compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala

```diff
@@ -353,8 +353,27 @@ class OrderingConstraint(private val boundsMap: ParamBounds,
     else {
       assert(contains(param1), i"$param1")
       assert(contains(param2), i"$param2")
-      val newUpper = param2 :: exclusiveUpper(param2, param1)
-      val newLower = param1 :: exclusiveLower(param1, param2)
+      // Is `order` called during parameter unification?
+      val unifying = isLess(param2, param1)
+      val newUpper = {
+        val up = exclusiveUpper(param2, param1)
+        if unifying then
+          // Since param2 <:< param1 already holds now, filter out param1 to avoid adding
+          //   duplicated orderings.
+          param2 :: up.filterNot(_ eq param1)
+        else
+          param2 :: up
+      }
+      val newLower = {
+        val lower = exclusiveLower(param1, param2)
+        if unifying then
+          // Do not add bounds for param1 since it will be unified to param2 soon.
+          // And, similarly filter out param2 from lowerly-ordered parameters
+          //   to avoid duplicated orderings.
```

Comment:
Are these changes correctness fixes or are they providing some other value (e.g. debug clarity or performance)?  On the face of it "avoid duplicated orderings" and "will be unified to param2 soon" sounds like the kind of optional things that doesn't impact the end result - but I'm wondering if that's right or not.

Reply:
Sorry for the late reply!

The special handling here actually avoid two different types of redundancy.

The first redundancy is that you may have duplicated entries in `lowerMap` and `upperMap` (which store the ordering between parameters) of `OrderingConstraint`. For example, you may get `A <: B, B` stored in the constriant.

This type of redundancy, actually, looks harmless. Actually, since the `lowerMap` and `upperMap` are all lists but not set, such situation will happen whenever some ordering is propagated along multiple different paths (and irrelated to the code change proposed by the PR). This will not bring erroneous behaviors of the compiler.

The second redundacy is that the constraint may end up storing `B <: A` even if B is unified to A. For example, we may have the constraint state as following.
```
Bounds:
  B  := A
Ordering:
  B  <: A
```

Such redundancy will break a patmat test. I could not figure out the exact reason, but preventing the redundancy ordering can solve the problem.

I am not sure if the broken patmat test is an expected behavior, since having the ordering `B <: A` while also having `B` unified to `A` will leave the constraints in a weird state. Anyway, I think I should refine the comments to clarify the issues here.